overlays/haskell.nix

```diff
@@ -256,49 +239,206 @@ final: prev: {
+              # E.g. after prepopulating CABAL_DIR/packages/hackage.haskell.org from the nix path
+              # /nix/store/ff4pn0yva7ndsrg2zshy8qxzlrfsr4cl-cabal-bootstrap-index-tarball-hackage.haskell.org/
+              # the configuration we want to write is still
+              #
+              #   repository hackage.haskell.org
+              #     url: http://hackage.haskell.org
+              #     secure: True
+              #
+              # and not
+              #
+              #   repository hackage.haskell.org
+              #     url: file:/nix/store/ff4pn0yva7ndsrg2zshy8qxzlrfsr4cl-cabal-bootstrap-index-tarball-hackage.haskell.org/
+              #     secure: True
+              #
+              # Doing this correctly is important for few reasons:
+              #
+              # 1. cabal as called by haskell.nix will produce exactly the same install plan as cabal
+              #    called outside of haskell.nix (other things equal, of course)
+              # 2. The url of the repositories will be visible in the install plan. Having the correct
+              #    urls there will allow us to know from where to fetch the packages tarballs at build
+              #    time.
+              # 3. We don't want to leak the nix path of the index into the derivation of the component
+              #    builder since this will cause unnecesary recompilation. In other words, the recipe to
+              #    compile a package has to only depend on its content, not on where the recipe is from
+              #    or how it is obtained.
+              #
+              # Using the correct url for the repository is only possible if the url is well-known. Again
+              # there is a difference in this regard among the different kind of repositories.
+              #
+              # - hackage: hackage's url is always well known.
+              # - extra-hackage-repos: since the user is responsible for downloading these repositories,
+              #   we don't really know where they are from, BUT, we also don't have to include them in
+              #   cabal.config because the user has already included them in the project configuration.
+              # - extra-hackage-tarballs: this is just a tarball passed to haskell.nix, we have no idea
+              #   if the corresponding repository is published anywhere. The best we can do in this case
+              #   is to form a url from the key of the extra-hackage-tarball and leave the user to decide.
+              #   The user could use a key corresponding to a real reository domain or overwrite the
+              #   packages source manually.
+              #   E.g. passing
+              #
+              #     extra-hackage-tarballs = { extra-hackage-demo = ./01-index.tar.gz; };
+              #
+              #   will produce in cabal.config
+              #
+              #     repository extra-hackage-demo
+              #       url: http://extra-hackage-demo/
```

Comment:
hmmmm. What about using the key _exactly_, like we do for e.g. `inputsMap`? So you would have to put `extra-hackage-tarballs = { "http://extra-hackage-demo" = ... }`

Reply:
Well, that is also reasonable but I don't know what the expectation is and I didn't want to alter it :) Note that `url` has to be a parsable url (I tried to just do `url: ${name}` so that users could add the `http://` on their end later on but it didn't work).